packages/eslint-plugin-material-ui/src/rules/disallow-react-api-in-server-components.js

```diff
@@ -1,43 +1,56 @@
-module.exports = {
+/// @ts-check
+
+const REACT_CLIENT_APIS = new Set([
+  'useState',
+  'useEffect',
+  'useLayoutEffect',
+  'useReducer',
+  'useTransition',
+  'createContext',
+]);
+
+/**
+ * @param {import('eslint').AST.Program} ast
+ * @param {string} directive
+ * @returns
+ */
+function hasDirective(ast, directive) {
+  return ast.body.some(
+    (statement) =>
+      statement.type === 'ExpressionStatement' &&
+      statement.expression.type === 'Literal' &&
+      statement.expression.value === directive,
+  );
+}
+
+module.exports = /** @type {import('eslint').Rule.RuleModule} */ ({
   create(context) {
     let hasUseClientDirective = false;
-    const apis = new Set([
-      'useState',
-      'useEffect',
-      'useLayoutEffect',
-      'useReducer',
-      'useTransition',
-      'createContext',
-    ]);
+    let hasUseServerDirective = false;
     return {
+      /** @param {import('eslint').AST.Program} node */
       Program(node) {
-        hasUseClientDirective = node.body.some(
-          (statement) =>
-            statement.type === 'ExpressionStatement' &&
-            statement.expression.type === 'Literal' &&
-            statement.expression.value === 'use client',
-        );
+        hasUseServerDirective = hasDirective(node, 'use server');
+        hasUseClientDirective = hasDirective(node, 'use client');
       },
       CallExpression(node) {
         if (
           !hasUseClientDirective &&
```

Comment:
Shouldn't this check for `!hasUseServerDirective` as well? 🤔

Reply:
Not, the marked APIs are not allowed in server components.